packages/core/src/State.ts

```diff
@@ -163,6 +190,14 @@ export class State<
     this.children = config.children;
 
     this.value = getStateValue(machine.root, this.configuration);
+    this.done = this.configuration.length
+      ? isInFinalState(this.configuration)
+      : false;
+    this.output =
+      config.output ??
+      (this.done
+        ? getOutput(this.configuration, this.context, this._event)
+        : undefined);
```

Comment:
This actually can't be computed in the constructor right now because it might lead to the `data` functions being called multiple times and thus this test fails:
https://github.com/statelyai/xstate/blob/10e07cf4c27913a1b9e49916fb76143fad8bb5e7/packages/core/test/final.test.ts#L198-L219

This happens because we `.clone()` the state a lot while transitioning.

This further tells me that perhaps we shouldn't be computing this stuff here, in this class, but rather just provide this as an argument.